packages/remix-react/components.tsx

```diff
@@ -712,52 +714,103 @@ export function Meta() {
         typeof routeModule.meta === "function"
           ? routeModule.meta({ data, parentsData, params, location })
           : routeModule.meta;
-      Object.assign(meta, routeMeta);
+      if (routeMeta) {
+        processMeta(meta, routeMeta);
+      }
     }
 
     parentsData[routeId] = data;
   }
 
   return (
     <>
-      {Object.entries(meta).map(([name, value]) => {
-        if (!value) {
-          return null;
+      {Array.from(meta.entries()).map(([key, descriptor]) => {
+        let { name, property, content } = descriptor;
+        if (key === "title" && typeof content === "string") {
+          return <title key={key}>{content}</title>;
         }
-
-        if (["charset", "charSet"].includes(name)) {
-          return <meta key="charset" charSet={value as string} />;
+        if (key === "charset" && typeof content === "string") {
+          return <meta key={key} charSet={content} />;
         }
-
-        if (name === "title") {
-          return <title key="title">{String(value)}</title>;
+        if (property !== undefined) {
+          return <meta key={key} property={property} content={content} />;
         }
-
-        // Open Graph tags use the `property` attribute, while other meta tags
-        // use `name`. See https://ogp.me/
-        let isOpenGraphTag = name.startsWith("og:");
-        return [value].flat().map((content) => {
-          if (isOpenGraphTag) {
-            return (
-              <meta
-                property={name}
-                content={content as string}
-                key={name + content}
-              />
-            );
-          }
-
-          if (typeof content === "string") {
-            return <meta name={name} content={content} key={name + content} />;
-          }
-
-          return <meta key={name + JSON.stringify(content)} {...content} />;
-        });
+        if (name !== undefined) {
+          return <meta key={key} name={name} content={content} />;
+        }
+        return <meta key={key} {...descriptor} />;
       })}
     </>
   );
 }
 
+/*
+ * This function processes a meta descriptor and adds it to the meta array.
+ * This converts the original object syntax to new array syntax.
+ */
+export function processMeta(
+  meta: MetaMap,
+  routeMeta: HtmlMetaDescriptor | HtmlMetaDescriptor[]
+) {
+  // normalize routeMeta to array format
+  let items: HtmlMetaDescriptor[] = Array.isArray(routeMeta)
+    ? routeMeta.map((item) =>
+        item.title
+          ? { key: "title", content: item.title }
+          : { key: generateKey(item), ...item }
+      )
+    : Object.entries(routeMeta)
+        .map(([key, value]) => {
+          if (!value) return [];
+
+          let propertyName = key.startsWith("og:") ? "property" : "name";
```

Comment:
This will need to be updated. See https://github.com/remix-run/remix/pull/4445

Reply:
Ok, I'll review this. Basically the key should be either `name` or `property` value. The only time you need an explicit `key` property is if you have duplicates, or there is no name/property like `<meta http-equiv="refresh" />`